nemo/collections/nlp/models/language_modeling/megatron_bert_model.py

```diff
@@ -112,25 +164,58 @@ def forward(self, input_ids, attention_mask, token_type_ids, lm_labels=None):
         return output_tensor
 
     def training_step(self, batch, batch_idx):
-        tokens, types, sentence_order, loss_mask, lm_labels, padding_mask = self.process_batch(batch)
-        if not self.cfg.bert_binary_head:
-            types = None
-        output_tensor = self(tokens, padding_mask, token_type_ids=types, lm_labels=lm_labels)
-        loss_dict = self.loss_func(loss_mask, sentence_order, output_tensor)
-        if 'sop loss' in loss_dict:
-            lm_loss = loss_dict['lm loss']
-            sop_loss = loss_dict['sop loss']
-            loss = lm_loss + sop_loss
-            reduced_loss = average_losses_across_data_parallel_group([loss, lm_loss, sop_loss])
-            self._reduced_loss_buffer.append(reduced_loss[0])
-            self._reduced_lm_loss_buffer.append(reduced_loss[1])
-            self._reduced_sop_loss_buffer.append(reduced_loss[2])
+
+        self._optimizer.zero_grad()  # DO WE NEED THIS
+        batch_for_pipeline = self.process_batch(batch)
+        tensor_shape = [self.cfg.encoder_seq_length, self.cfg.micro_batch_size, self.cfg.hidden_size]
+
+        if self.megatron_amp_o2 and not self.cfg.get('sequence_parallel', False):
+            custom_sync_context_handler = self._optimizer.no_sync
+        else:
+            # TODO: enable async grad all reduce for O1/autocast mixed precision training
+            custom_sync_context_handler = None
+
+        losses_reduced_per_micro_batch = forward_backward_no_pipelining(
+            forward_step_func=self.get_forward_output_and_loss_func(),
+            batch=batch_for_pipeline,
+            model=self.model,
+            forward_only=False,
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+            grad_scaler=self.trainer.precision_plugin.scaler if self.cfg.precision == 16 else None,
+            custom_sync_context_handler=None,
+        )
+
+        loss_tensors_list = [loss_reduced['avg'] for loss_reduced in losses_reduced_per_micro_batch]
+        loss_tensor = torch.vstack(loss_tensors_list)
+        loss_mean = loss_tensor.mean(axis=0)
```

Comment:
The loss_mean doesn't seem to be averaged across data parallel group. Could you confirm where it is averaged?

Reply:
It happens in the loss_func, inside get_forward_output_and_loss_func